src/Microsoft.VisualStudio.Editor.Razor/BraceSmartIndenter.cs

```diff
@@ -0,0 +1,218 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Text;
+using Microsoft.AspNetCore.Razor.Language.Legacy;
+using Microsoft.CodeAnalysis.Razor;
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.Text.Editor;
+using Microsoft.VisualStudio.Text.Operations;
+using ITextBuffer = Microsoft.VisualStudio.Text.ITextBuffer;
+
+namespace Microsoft.VisualStudio.Editor.Razor
+{
+    /// <summary>
+    /// This class is responsible for handling situations where Roslyn and the HTML editor cannot auto-indent Razor code.
+    /// </summary>
+    /// <example>
+    /// Attempting to insert a newline (pipe indicates the cursor):
+    /// @{ |}
+    /// Should result in the text buffer looking like the following:
+    /// @{
+    ///     |
+    /// }
+    /// This is also true for directive block scenarios.
+    /// </example>
+    internal class BraceSmartIndenter : IDisposable
```

Comment:
`IDisposable` for now since the `VisualStudioRazorParser` is. This will probably change once we own the lifetime of the parser (my next item).